### PR TITLE
Persisting Vote instead of Candidate, real and fake keys displayed on separate JSP

### PR DESCRIPTION
**Pull request by kelseyball**

### CODE REVIEW

**Comment by kelseyball** [comment]
Sorry I did way too many things on this branch. Features to examine:

1. Now we're objectifying/persisting a Vote, which has a candidate (an index into the Candidate enum) and boolean indicating if it's been published or not. This is because Objectify doesn't like persisting enums.
2. castvote.jsp now dynamically populates the radio buttons using the enum, instead of hard-coding the candidate choices.
3. A vote is processed (persisted) in the castVoteServlet, which sends the vote information (for the real and fakes) to displayVoteKeys.jsp to be displayed
4. displayVoteKeys.jsp now prompts the user to see more fake votes. A separate servlet (castFakeVoteServlet) and jsp (displayFakes) cast and display additional batches of fake votes.
5. If the user doesn't want to see more fake batches, they're redirected to goodbye (added a GoodbyeServlet for this redirect, where we could also do any cleanup if needed).

**Comment by kelseyball** [comment]
6. There's a numFakeVoteBatches counter that's passed back and forth between the castFakeVoteServlet and displayFakes.jsp. We'll want to persist this and use it for the CronServlet tallies/condition checking, but haven't thought about best way to do that, objectify-wise.

**Comment by dylanzika** [comment]
Shouldn't that equation still be solvable ? That we derived and wrote down in our block diagram ?

Sincerely,

Dylan Zika

**Comment by dylanzika** [comment]
What if we always send out a exact number of fake votes but only let the person voting have a subset of the fake votes. Don't let them ask for 100 sets. Give them atleast one vote per candidate but always send out 100 fake votes to the system ? 

Sincerely,

Dylan Zika

File: OpenVote/src/com/openvote/CastVoteServlet.java [hunk 1]
```diff
@@ -1,54 +1,53 @@
...
     	
-		resp.setContentType("text/plain");
-		resp.getWriter().println("You voted for " + candidate_str + ".");
-		resp.getWriter().println("Your key is " + vote.getVoteKey().toString() + ".");
-		
+    	// cast vote for all fake candidates
+    	ArrayList<Vote> fakeVotes = new ArrayList<Vote>();
+		for (Candidate c: Candidate.values()) {
+			Vote fakeVote = new Vote(c.ordinal());
+    		ofy().save().entity(fakeVote).now();
```
Comment (williammills) [inline OpenVote/src/com/openvote/CastVoteServlet.java]: Style: tab please :)

File: OpenVote/src/com/openvote/CastVoteServlet.java [hunk 2]
```diff
@@ -1,54 +1,53 @@
...
+    		fakeVotes.add(fakeVote);
```
Comment (williammills) [inline OpenVote/src/com/openvote/CastVoteServlet.java]: Style: tab please :)

File: OpenVote/src/com/openvote/CastVoteServlet.java [hunk 3]
```diff
@@ -1,54 +1,53 @@
...
+    	ofy().save().entity(vote).now(); 	
+    	//TODO: verify that the autogenerated id is unique
     	
-		resp.setContentType("text/plain");
-		resp.getWriter().println("You voted for " + candidate_str + ".");
-		resp.getWriter().println("Your key is " + vote.getVoteKey().toString() + ".");
-		
+    	// cast vote for all fake candidates
+    	ArrayList<Vote> fakeVotes = new ArrayList<Vote>();
+		for (Candidate c: Candidate.values()) {
```
Comment (williammills) [inline OpenVote/src/com/openvote/CastVoteServlet.java]: We should put this loop into a private method so it isn't clear how the list is built. Also, I think we should have the real vote in this list (done in the private method). That way, it's indistinguishable from the fake votes

File: OpenVote/src/com/openvote/CastVoteServlet.java [hunk 4]
```diff
@@ -1,54 +1,53 @@
...
 	}
 	
     public void doPost(HttpServletRequest req, HttpServletResponse resp)
             throws IOException {
     	String candidate_str = req.getParameter("candidate");
-    	VoteKey vk = new VoteKey(UUID.randomUUID()); //TODO this needs to be checked for uniqueness
-    	Vote vote = new Vote(vk, Candidate.valueOf(candidate_str));
     	
+    	// cast real vote
+    	Vote vote = new Vote(Candidate.valueOf(candidate_str).ordinal());
```
Comment (williammills) [inline OpenVote/src/com/openvote/CastVoteServlet.java]: Should we put this in the list of fake votes to treat them all the same?

File: OpenVote/src/com/openvote/CastVoteServlet.java [hunk 5]
```diff
@@ -1,54 +1,53 @@
...
+    	// cast vote for all fake candidates
+    	ArrayList<Vote> fakeVotes = new ArrayList<Vote>();
+		for (Candidate c: Candidate.values()) {
+			Vote fakeVote = new Vote(c.ordinal());
+    		ofy().save().entity(fakeVote).now();
+    		fakeVotes.add(fakeVote);
+    	}
+    	
+    	// pass real and fake vote information to display keys page
+		req.setAttribute("realVote", vote);
```
Comment (williammills) [inline OpenVote/src/com/openvote/CastVoteServlet.java]: We can leave this out, and when we want to display things to people just tell them "please check that your indended candidate has one more vote than all others."

File: OpenVote/src/com/openvote/GoodbyeServlet.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+public class GoodbyeServlet extends HttpServlet {
+	
+
+    public void doPost(HttpServletRequest req, HttpServletResponse resp)
+            throws IOException {
+    	
+    	// any clean-up required here?
+    	
+    	resp.sendRedirect("/goodbye.jsp");
```
Comment (williammills) [inline OpenVote/src/com/openvote/GoodbyeServlet.java]: If this is all the servlet does, I don't think we need it. We can just have a simple redirect whenever we're going to goodbye

File: OpenVote/src/com/openvote/LoginServlet.java [hunk 7]
```diff
@@ -10,11 +10,16 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-import java.io.IOException;
+
 import javax.servlet.http.*;
+import com.googlecode.objectify.ObjectifyService;
 
 @SuppressWarnings("serial")
 public class LoginServlet extends HttpServlet {
```
Comment (williammills) [inline OpenVote/src/com/openvote/LoginServlet.java]: Wait, why do we have this? Aren't we using Google login?

**Pull request merged**